### PR TITLE
file creation

### PR DESCRIPTION
**Pull request by Himanshu-Singh-Chauhan**
use @tree and select `File Tree` context provider to send you project directory structure to ai.

new file can be created from chat now, a chip will appear whenever ai suggests creating a new file.

### CODE REVIEW

File: core/context/providers/FileTreeContextProvider.ts [hunk 1]
```diff
@@ -74,9 +74,17 @@ class FileTreeContextProvider extends BaseContextProvider {
...
+        this way my code editor will parse it and create the file for me.
+        `.trim();
+
     return [
       {
-        content: `Here is a file tree of the current workspace:\n\n${trees.join(
+        content: fileCreationPrompt + `\n\nHere is a file tree of the current workspace:\n\n${trees.join(
           "\n\n",
         )}`,
         name: "File Tree",
```
Comment (Fryingpannn) [inline core/context/providers/FileTreeContextProvider.ts:90]: I see so there was an existing file tree context provider

Comment (Himanshu-Singh-Chauhan) [reply core/context/providers/FileTreeContextProvider.ts:90]: haha, yeah, i found this after i almost made similar to this, then i discarded mine 🥲

File: core/context/providers/FileTreeContextProvider.ts [hunk 2]
```diff
@@ -74,9 +74,17 @@ class FileTreeContextProvider extends BaseContextProvider {
       trees.push(formatFileTree(subDirTree));
     }
 
+
+    const fileCreationPrompt = `Whenever you respond contains some suggestions or instructions to create new files,
```
Comment (Fryingpannn) [inline core/context/providers/FileTreeContextProvider.ts:78]:
```suggestion
    const fileCreationPrompt = `Whenever your response contains any suggestions or instructions to create new files, 
```

File: core/config/default.ts [hunk 3]
```diff
@@ -33,6 +33,7 @@ export const FREE_TRIAL_MODELS: ModelDescription[] = [
 ];
 
 export const defaultContextProvidersVsCode: ContextProviderWithParams[] = [
+  { name: "tree", params: {} },
```
Comment (Fryingpannn) [inline core/config/default.ts:36]:
```suggestion
  { name: "directory-structure", params: {} },
```

File: core/protocol/core.ts [hunk 4]
```diff
@@ -132,6 +132,7 @@ export type ToCoreFromIdeOrWebviewProtocol = {
   "index/setPaused": [boolean, void];
   "index/forceReIndex": [undefined | string, void];
   "index/indexingProgressBarInitialized": [undefined, void];
+  "fileExists2": [filePath: string, boolean];
```
Comment (Fryingpannn) [inline core/protocol/core.ts:135]: curious as to why it's called fileExists2 (why the 2)

File: core/context/providers/FileTreeContextProvider.ts (hunk 1 above)
Comment (Fryingpannn) [inline core/context/providers/FileTreeContextProvider.ts:90]: I think we should rename "@ File Tree" to "@ Directory Structure", more clear to the user. As a first time user, I would also be searching for directory more so than tree.

File: gui/src/components/markdown/FileCreateChip.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        <span
+          className="select-none rounded-md text-center cursor-pointer"
+          onClick={handleCreateFile}
+          style={{ display: "flex", alignItems: "center" }} // Ensure flexbox centering
+        >
+          {fileExists ? (
+            <CheckCircleIcon className="w-4 h-4" strokeWidth={2} color="green" />
+          ) : (
+            <PlusIcon className="w-4 h-4" strokeWidth={2} />
+          )}
```
Comment (Fryingpannn) [inline gui/src/components/markdown/FileCreateChip.tsx:87]: I think for the UI of this, aside from your mention of the file icon on the left, we should also add text on the right side to indicate either "Create file" or "Open file". This small text would be part of the "+" button, and together they would look like a clickable button. We want to make it super easier for the user.

image

**Review by Fryingpannn — DISMISSED** [review]

File: core/context/providers/FileTreeContextProvider.ts (hunk 2 above)
Comment (Himanshu-Singh-Chauhan) [reply core/context/providers/FileTreeContextProvider.ts]: renamed tree to directory

File: core/config/default.ts (hunk 3 above)
Comment (Himanshu-Singh-Chauhan) [reply core/config/default.ts]: renamed tree to directory

File: core/protocol/core.ts (hunk 4 above)
Comment (Himanshu-Singh-Chauhan) [reply core/protocol/core.ts]: this was for testing, removed this

File: core/context/providers/FileTreeContextProvider.ts (hunk 1 above)
Comment (Himanshu-Singh-Chauhan) [reply core/context/providers/FileTreeContextProvider.ts]: renamed tree to directory

File: gui/src/components/markdown/FileCreateChip.tsx (hunk 5 above)
Comment (Himanshu-Singh-Chauhan) [reply gui/src/components/markdown/FileCreateChip.tsx:129]: improved some ui, fixed on solarized, we can play around with colors more, but the one I have set now plays nice with lots of themes.

**Review by Fryingpannn — COMMENTED** [review]
Need some responsiveness 👀
When too long, the create button is cutoff.

image

**Comment by Fryingpannn** [comment]
Anotha test case failed D:
Prompt:
```
 Currently when someone is subscribed to monthly and they want to upgrade to annual, they must wait until their subscription ends.

What we want:

A monthly subscriber clicks a button called "Upgrade to annual" on the dashboard, which upgrades their subscription to the annual plan at the end of their current period.

Info

Need to check Stripe API to see how to change subscription
```

image

**Comment by nang-dev** [comment]
Screenshot 2024-09-17 at 5 18 48 PM

think im missing something here when trying to run

**Comment by nang-dev** [comment]
tbh might be something on my end

**Comment by nang-dev** [comment]
hmm works when i checkout main, maybe ill wait to tallk to pan or him

File: gui/src/components/mainInput/resolveInput.ts [hunk 6]
```diff
@@ -114,6 +116,42 @@ async function resolveEditorContent(
...
+      selectedCode,
+    },
+  );
+
+  console.dir("comparing previous and current")
+  // console.dir((whatinit as any).state.directoryItems)
+  console.dir(previousDirectoryItems !== directoryItems[0].content)
+  
+  if (previousDirectoryItems !== directoryItems[0].content ) {
+    console.dir("previousDirectoryItems ---XXXXXXXXXXX")
```
Comment (nang-dev) [inline gui/src/components/mainInput/resolveInput.ts]: remove too

**Review by nang-dev — APPROVED** [review]

**Pull request merged**